Add Elasticsearch query input plugin

lpic10 opened the pull request:
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

This implements #2506 

I've been using something similar to this code for some time already, I finally decided to clean it up a bit and add some tests and documentation.

lpic10 commented:
Some to-do tasks in my list: 

- [ ] Support for Elasticsearch 6.x
- [ ] Support for percentiles aggregation

glinton commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,287 @@
+  # password = "mypassword"
+
+  ## Optional SSL Config
+  # ssl_ca = "/etc/telegraf/ca.pem"
```
Use `tls_[ca|cert|key]`

glinton commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,287 @@
+	Timeout             internal.Duration
+	HealthCheckInterval internal.Duration
+	Aggregations        []Aggregation `toml:"aggregation"`
+	SSLCA               string        `toml:"ssl_ca"`   // Path to CA file
```
No need for these SSL/TLS attributes if you use `./internal/tls` and the shared tls config [like apache uses](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/apache/apache.go#L96)

glinton commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,287 @@
+
+	var clientOptions []elastic.ClientOptionFunc
+
+	tlsCfg, err := internal.GetTLSConfig(e.SSLCert, e.SSLKey, e.SSLCA, e.InsecureSkipVerify)
```
Use `internal/tls` and `TlsConfig()`

glinton commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,287 @@
+	if err != nil {
+		return err
+	}
+	tr := &http.Transport{
```
Make the transport and http.Client a member of `ElasticsearchQuery`, again similar to [the apache plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/apache/apache.go#L69)

lpic10 commented:
Thanks @glinton for the initial review, I've implemented the changes concerning TLS, the sample config and the README with examples. 

In addition I removed some commented code, and I changed the check for version to allow it to work on newer ES (6.x,) which I run the tests against and it seems to be working fine..

gl1ridae commented:
Hi! Is there any chanse this MR will be merged in the next version? Any additional help required on this?
@danielnelson

ddemin commented:
Really need this too =)

vishrantgupta commented:
Hi @lpic10 @glinton, any plans when it can be merged?

srebhan commented:
@lpic10 are you still interested in getting this merged?

hexinal commented:
@srebhan 
not op, but I maintain a "fork" of this in production so there is at least one person who would be interested in getting this in upstream :smile: . 
Let me know if I can help.

lpic10 commented:
> @lpic10 are you still interested in getting this merged?

Yes, even if I'm not actively using this PR as my job attributions have changed since.
I will try to have a look at it during the weekend.

srebhan commented:
@hexinal please coordinate with @lpic10. You will be definitively required for testing... ;-)

srebhan commented on plugins/inputs/elasticsearch_query/README.md:
````diff
@@ -0,0 +1,137 @@
+  ## Set the interval to check if the Elasticsearch nodes are available
+  ## Setting to "0s" will disable the health check (not recommended in production)
+  health_check_interval = "10s"
+  ## HTTP basic authentication details (eg. when using x-pack)
+  # username = "telegraf"
+  # password = "mypassword"
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+
+[[inputs.elasticsearch_query.aggregation]]
+  measurement_name = "measurement"
+  index = "index-*"
+  date_field = "@timestamp"
+  ## Time window to query (eg. "1m" to query documents from last minute).
+  ## Normally should be set to same as collection interval
+  query_period = "1m"
+
+  ## Optional parameters:
+  ## Lucene query to filter results
+  filter_query = "*"
+  ## Fields to aggregate values (must be numeric fields)
+  metric_fields = ["metric"]
+  ## Aggregation function to use on the metric fields
+  ## Valid values are: avg, sum, min, max, sum
+  metric_function = "avg"
+  ## Fields to be used as tags
+  ## Must be non-analyzed fields, aggregations are performed per tag
+  tags = ["field.keyword", "field2.keyword"]
+  ## Set to true to not ignore documents when the tag(s) above are missing
+  include_missing_tag = true
+  ## String value of the tag when the tag does not exist
+  ## Used when include_missing_tag is true
+  missing_tag_value = "null"
+
+```
````
Could you please make this a minimal useful version of the config. That is leave out as many options as possible and just keep what _has to_ be changed by the user. This way it is easier to for new users to start.

srebhan commented on plugins/inputs/elasticsearch_query/README.md:
````diff
@@ -0,0 +1,137 @@
+### Search the average response time, per URI and per response status code
+
+```toml
+[[inputs.elasticsearch_query.aggregation]]
````
Maybe you could add a not that the `[[inputs.elasticsearch_query]]` section is still required. :-)

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+package elasticsearch_query
+
+import elastic "gopkg.in/olivere/elastic.v5"
```
Would it make sense to use the official client at https://github.com/elastic/go-elasticsearch?

lpic10 replied:
That would require a complete rewrite of the code I'm afraid.
Also this client is the same already in use on elasticsearch output plugin.

The fact that it is using a rather "old" (v5) version maybe is an issue or will become one soon.
One option would be to import multiple versions (v5, v6, v7...) and use the right client according to the version of the cluster detected.

srebhan replied:
Fair enough. Let's start with this dependency then and maybe switch to the other client with higher versions...

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+	var measurements = make(map[string][]*aggKey)
+
+	for _, aggregationQuery := range *aggregationQueryList {
+		m := measurements[aggregationQuery.measurement]
+		a := aggKey{function: aggregationQuery.function, field: aggregationQuery.field, name: aggregationQuery.name}
+
+		if m == nil {
+			m = []*aggKey{&a}
+		} else {
+			m = append(m, &a)
+		}
+		measurements[aggregationQuery.measurement] = m
```
How about
```suggestion
		a := aggKey{function: aggregationQuery.function, field: aggregationQuery.field, name: aggregationQuery.name}

		if m := measurements[aggregationQuery.measurement]; m != nil {
			measurements[aggregationQuery.measurement] = append(m, &a)
		} else {
			measurements[aggregationQuery.measurement] = []*aggKey{&a}
		}
```

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+
+	for _, aggKey := range aggKeys {
+		resp := e.getResponseAggregation(aggKey.function, aggKey.name, aggs)
+
+		if resp != nil {
+
```
Please do
```suggestion

		if resp == nil {
			// Maybe do some logging here if this is an error...
			continue
		}

```

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+			case *elastic.AggregationValueMetric:
+
+				// we've found a metric aggregation, add to field map
+				fields[aggKey.name] = float64(0)
+
+				if resp.Value != nil {
+					fields[aggKey.name] = *resp.Value
+				}
```
Here you might use an else clause to not set the value twice.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+
+			case *elastic.AggregationBucketKeyItems:
+				// we've found a terms aggregation, iterate over the buckets and try to retrieve the inner aggregation values
+				for item := range resp.Buckets {
```
Don't you get a bucket when iterating over the buckets ([reference](https://github.com/olivere/elastic/blob/be977d75416e9a352a0853165c9d1786e3d43c51/search_aggs.go#L1251))?

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+					bucket := resp.Buckets[item]
+
+					fields["doc_count"] = bucket.DocCount
+					tags[aggKey.name] = bucket.Key.(string)
```
Please check this type assertion as otherwise telegraf will panic when we encounter a non-string key.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+		fields := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		err := e.recurseResponse(aggKey, searchResult.Aggregations, measurementName, fields, tags)
```
Please put a comment here that you need to recursively resolve the fields and tags as otherwise it's strange to pass the data-structures down without using them afterwards.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+				return nil
+
+			case *elastic.AggregationPercentilesMetric:
+				// TODO
```
:-) So please do or delete the case. :-)

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+			case *elastic.AggregationPercentilesMetric:
+				// TODO
+				return nil
+			}
```
You should add a `default` section and return an error for cases where you cannot handle the type. This catches problems where types are added to the client but we forgot to add them here.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+		}
+	}
+
+	if len(fields) > 0 {
+		e.acc.AddFields(measurementName, fields, tags)
+	}
```
I think this one has to go one level up. We want to collect the fields and values for the measurement/metric here and then add the metric when complete in `parseAggregationResult()`.

lpic10 replied:
I think if we move up the metric may not be complete when adding to the Accumulator, I will re-check that

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+						return err
+					}
+				}
+				return nil
```
Is this one really correct? You will ignore all other responses after you've hit one of the `elastic.AggregationBucketKeyItems` type. That seems wrong.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,117 @@
+		agg, found = aggs.Max(aggName)
+	case "terms":
+		agg, found = aggs.Terms(aggName)
+		// TODO:
+		// case "percentile":
+		// 	agg, found = aggs.Percentiles(aggName)
```
Please remove the TODOs.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,184 @@
+func (e *ElasticsearchQuery) runAggregationQuery(ctx context.Context, aggregation Aggregation, aggregationQueryList []aggregationQueryData) (*elastic.SearchResult, error) {
+
+	now := time.Now().UTC()
+	from := now.Add(aggregation.QueryPeriod.Duration * -1)
```
```suggestion
	from := now.Sub(aggregation.QueryPeriod.Duration)
```

srebhan replied:
Huh? Where is this change?

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,184 @@
+	}
+	searchResult, err := search.Do(ctx)
+
+	if err != nil {
+		if searchResult != nil {
```
Please fuse those two ifs.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,184 @@
+	query = query.Filter(elastic.NewQueryStringQuery(filterQuery))
+	query = query.Filter(elastic.NewRangeQuery(aggregation.DateField).From(from).To(now))
+
+	search := e.ESClient.Search().
+		Index(aggregation.Index).
+		Query(query).
+		Size(0)
```
Put in one line please.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,184 @@
+		}
+
+		for _, index := range resp {
+			for _, _type := range index.(map[string]interface{})["mappings"].(map[string]interface{}) {
+				for _, field := range _type.(map[string]interface{}) {
```
Please check these type assertions!

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,184 @@
+		for _, index := range resp {
+			for _, _type := range index.(map[string]interface{})["mappings"].(map[string]interface{}) {
+				for _, field := range _type.(map[string]interface{}) {
+					fname := field.(map[string]interface{})["full_name"].(string)
+					for _, fieldType := range field.(map[string]interface{})["mapping"].(map[string]interface{}) {
+						ftype = fieldType.(map[string]interface{})["type"].(string)
```
Same here.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,184 @@
+	// case "percentile":
+	// 	agg = elastic.NewPercentilesAggregation().Field(aggfield)
+	default:
+		return agg, fmt.Errorf("Aggregation function %s not supported", function)
```
In case you error please return `nil` for the value if possible. This way you directly see that you cannot rely on the value and do not need to trace back the variable.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const description = `Queries Elasticseach`
```
Maybe "Derive metrics from aggregating Elasticsearch query results"?

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+  ## Multiple urls can be specified as part of the same cluster,
+  ## this means that only ONE of the urls will be written to each interval.
+  urls = [ "http://node1.es.example.com:9200" ] # required.
+  ## Elasticsearch client timeout, defaults to "5s" if not set.
```
Please add a blank line between option groups.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+  ## this means that only ONE of the urls will be written to each interval.
+  urls = [ "http://node1.es.example.com:9200" ] # required.
+  ## Elasticsearch client timeout, defaults to "5s" if not set.
+  timeout = "5s"
```
Please comment default settings. Use the default setting as value. This way the user sees what is used but also knows that the option is optional.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+  timeout = "5s"
+  ## Set to true to ask Elasticsearch a list of all cluster nodes,
+  ## thus it is not necessary to list all nodes in the urls config option
+  enable_sniffer = false
```
Maybe `cluster_node_discovery`? We do not really sniff do we?

lpic10 replied:
This setting is coming from the client: https://github.com/olivere/elastic/wiki/Sniffing
I'd be fine by removing or hiding this (and health check).

srebhan replied:
Yeah I'm only asking because the name is not very self-containing and it's barely expresses what it does. Furthermore, sniffing might have a negative co-notation... But if you want to keep it, it's up to you.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+  ## Normally should be set to same as collection interval
+  query_period = "1m"
+
+  ## Optional parameters:
+  ## Lucene query to filter results
+  filter_query = "*"
+  ## Fields to aggregate values (must be numeric fields)
+  metric_fields = ["metric"]
+  ## Aggregation function to use on the metric fields
+  ## Valid values are: avg, sum, min, max, sum
+  metric_function = "avg"
+  ## Fields to be used as tags
+  ## Must be non-analyzed fields, aggregations are performed per tag
+  tags = ["field.keyword", "field2.keyword"]
+  ## Set to true to not ignore documents when the tag(s) above are missing
+  include_missing_tag = true
+  ## String value of the tag when the tag does not exist
+  ## Used when include_missing_tag is true
+  missing_tag_value = "null"
```
Comment optional parameters in the form
```
## Some description
# parameter = "default value"
```

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+
+type ElasticsearchQuery struct {
+	URLs                []string `toml:"urls"`
+	Username            string
+	Password            string
+	EnableSniffer       bool
+	Tracelog            bool
+	Timeout             internal.Duration
+	HealthCheckInterval internal.Duration
```
Please add toml annotations. This eases debugging in case you are searching for the use of an option.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	Aggregations        []Aggregation `toml:"aggregation"`
+	tls.ClientConfig
+	httpclient *http.Client
+	ESClient   *elastic.Client
```
Do we really need to export this?

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	tls.ClientConfig
+	httpclient *http.Client
+	ESClient   *elastic.Client
+	acc        telegraf.Accumulator
```
Usually the accumulator is passed down the functions and not stored in the structure. If you do not have a striking reason please follow this unwritten rule. :-)

lpic10 replied:
I think I did that way because of the `recurseResponse` function. I thought it was better to have it "globally" than passing down each recursion. I will try to revisit that.

srebhan replied:
Ok that is a point, but maybe you can change the recursive function in a way that it returns what should be added to `acc` and not doing it while traversing the tree...

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	acc        telegraf.Accumulator
+}
+
+type Aggregation struct {
```
Can we make this struct private and if not can we use a more Elasticsearch-specific name!?

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+}
+
+type Aggregation struct {
+	Index             string
+	MeasurementName   string
+	FilterQuery       string
+	QueryPeriod       internal.Duration
+	MetricFields      []string `toml:"metric_fields"`
+	DateField         string
+	Tags              []string `toml:"tags"`
+	IncludeMissingTag bool
+	MissingTagValue   string
+	MetricFunction    string
```
Please also add toml tags.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	MetricFunction    string
+}
+
+type aggKey struct {
+	measurement string
+	name        string
+	function    string
+	field       string
+}
+
+type aggregationQueryData struct {
+	aggKey
+	isParent    bool
+	aggregation elastic.Aggregation
+}
```
Please move these definitions to one of the `aggregation_...` files as you don't use them here.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	return description
+}
+
+func (e *ElasticsearchQuery) init() error {
```
If you call this function `Init()` it will be automatically called by telegraf after the config parameters are parsed and set.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+		elastic.SetHealthcheckInterval(e.HealthCheckInterval.Duration),
+	)
+
+	if e.Username != "" && e.Password != "" {
```
This would prevent an empty password. If an empty password is valid use
```suggestion
	if e.Username != "" {
```
instead.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	)
+
+	if e.Username != "" && e.Password != "" {
+		clientOptions = append(clientOptions,
+			elastic.SetBasicAuth(e.Username, e.Password),
```
No line wrap here.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+		return fmt.Errorf("Elasticsearch urls is not defined")
+	}
+
+	err := e.connectToES()
+	return err
```
```suggestion
	return e.connectToES()
```

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+		)
+	}
+
+	if e.Tracelog {
+		clientOptions = append(clientOptions,
+			elastic.SetTraceLog(log.New(os.Stdout, "", log.LstdFlags)),
+		)
+	}
```
Where is this option set? Usually logging to stdout is an absolute exception. Can we remove this? Or at least guard with a config option even though I prefer the removal.

lpic10 replied:
That is for debugging the client, I agree it should not be in the release version as it pollutes the stdout and that's why I hid it from the config.

srebhan replied:
I see. However, I'd like this removed from the code that goes to telegraf. If possible and necessary add some debug-logging using the telegraf logger...

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	// quit if ES version is not supported
+	i, err := strconv.Atoi(strings.Split(esVersion, ".")[0])
```
Please check if the split returned anything, otherwise telegraf will panic here.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	// quit if ES version is not supported
+	i, err := strconv.Atoi(strings.Split(esVersion, ".")[0])
+	if err != nil || i < 5 {
+		return fmt.Errorf("Elasticsearch query: ES version not supported: %s", esVersion)
```
Errors have to start lower-case. Please also change the other instances. Furthermore it would be nice here to specify the supported version. Something like "ES versions < 5.x are not supported".

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+}
+
+func (e *ElasticsearchQuery) Gather(acc telegraf.Accumulator) error {
+	if err := e.init(); err != nil {
+		return err
+	}
```
Please rename the function above to `Init()` and remove this code.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+		wg.Add(1)
+		go func(agg Aggregation) {
+			defer wg.Done()
+			err := e.esAggregationQuery(agg)
```
Please pass the accumulator down here instead of storing it in the structure.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+	ctx, cancel := context.WithTimeout(context.Background(), e.Timeout.Duration)
+	defer cancel()
+
+	if aggregation.DateField == "" {
+		return fmt.Errorf("Field 'date_field' not set")
+	}
+
```
Maybe check this before constructing the context.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+		return fmt.Errorf("Field 'date_field' not set")
+	}
+
+	mapMetricFields, err := e.getMetricFields(ctx, aggregation)
+	if err != nil {
+		return err
+	}
+
+	aggregationQueryList, err := e.buildAggregationQuery(mapMetricFields, aggregation)
+	if err != nil {
+		return err
+	}
+
```
Do you really need to do this in every gather cycle or is it sufficient to do it once at the beginning and only set the timeinterval here?

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+
+func (e *ElasticsearchQuery) esAggregationQuery(aggregation Aggregation) error {
+
+	ctx, cancel := context.WithTimeout(context.Background(), e.Timeout.Duration)
```
This way the timeout is for both operations that query the server (`getMetricFields()` and `runAggreagtionQuery()`). Is this intentional?

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+		return err
+	}
+
+	if searchResult.Aggregations != nil {
+		err = e.parseAggregationResult(&aggregationQueryList, searchResult)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = e.parseSimpleResult(aggregation.MeasurementName, searchResult)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
```
You can shorten to
```suggestion
	if searchResult.Aggregations != nil {
		err = e.parseAggregationResult(&aggregationQueryList, searchResult)
	} else {
		err = e.parseSimpleResult(aggregation.MeasurementName, searchResult)
	}

	return err
```
But that's a matter of taste...

srebhan reviewed (changes requested):
First of all thank you very much for submitting this @lpic10! Here comes a quick first round...

Please check the `recurseResponse()` function thoroughly as it looks fishy to me. That being said, it would be nice if you can come up with some test-cases that solely test the parsing part (`recurseResponse()` and friends) without the need of a server. My experience is that this are usually the tricky parts that break on changes.

Furthermore, make use of the `Init()` function that is automatically called by telegraf once after passing in the config variables. Use this function to prepare as much of your queries as possible and to validate user settings.

As an idea rather than a hard requirement I want to suggest to remember the last query time and dynamically adapt the aggregation window to the time between last run and now. This should be the default IMO and there should be the possibility to override this by the `query_period`.

Looking forward to the next round!

lpic10 commented:
> First of all thank you very much for submitting this @lpic10! Here comes a quick first round...

Thanks for the review, very much needed!

I think I touched almost everything you mentioned. Except maybe implementing tests on the `recurseResponse` without the use of a server. 

This contribution was done few years ago, and because of that it is not supporting latest releases of Elasticsearch (>= 7.x)

Currently telegraf is using the elasticsearch version 7.2.0 for the integration tests, and this won't work with this plugin.
Maybe it is time already to upgrade the go elastic client to v7 for this (and the output plugin).

Another option would be to import v5, v6 and v7 client versions and use the appropriate one depending on the version of the cluster. This could be set via config parameter, or autodetected.

> As an idea rather than a hard requirement I want to suggest to remember the last query time and dynamically adapt the aggregation window to the time between last run and now. This should be the default IMO and there should be the possibility to override this by the `query_period`.

Is there any example of plugin using such mechanism?

> Looking forward to the next round!

srebhan commented:
> Currently telegraf is using the elasticsearch version 7.2.0 for the integration tests, and this won't work with this plugin.
Maybe it is time already to upgrade the go elastic client to v7 for this (and the output plugin).
Another option would be to import v5, v6 and v7 client versions and use the appropriate one depending on the version of the cluster. This could be set via config parameter, or autodetected.

This depends a bit on what you want. The approach using multiple client versions is preferable IMO as it allows to keep up with ElasticSearch releases in the future. But I'm also happy if you want to get this in first and "just" prepare later addition of v6 and v7.

Regarding the tests you could setup a `httptest` server (see `http/http_test.go` for an example) and return a prerecorded response for testing. This would eliminate the requirement for a real ES instance (keep this for long-tests :-)).

> Is there any example of plugin using such mechanism?
You might check `sysstat/sysstat.go`. It's just you remember the time of the last collection and use the time difference. However, please keep the user-option to override that mechanism... I guess I have to submit a PR to let plugins know the actually configured telegraf interval...

srebhan commented on .vscode/launch.json:
```diff
@@ -0,0 +1,19 @@
+{
```
Please remove this file. :-)

lpic10 replied:
ops, sorry, not sure how that ended up there :)

srebhan commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Same here. Please remove the `.vscode` subfolder and maybe add it to `gitignore`.

srebhan commented on plugins/inputs/elasticsearch_query/README.md:
```diff
@@ -0,0 +1,98 @@
+  # ssl_key = "/etc/telegraf/key.pem"
+  ## Use SSL but skip chain & host verification
+  # insecure_skip_verify = false
+
```
Please close this as resolved.

srebhan commented on plugins/inputs/elasticsearch_query/README.md:
````diff
@@ -0,0 +1,156 @@
+```
+
+## Examples
+
````
Please add
```suggestion
Please note that the `[[inputs.elasticsearch_query]]` is still required for all of the examples below.
```

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,154 @@
+	"fmt"
+
+	"github.com/influxdata/telegraf"
+	elastic "gopkg.in/olivere/elastic.v5"
```
As per our discussion, please prepare for later versions by
```suggestion
	elastic5 "gopkg.in/olivere/elastic.v5"
```
or similar.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (e *ElasticsearchQuery) parseSimpleResult(measurement string, searchResult *elastic.SearchResult, acc telegraf.Accumulator) error {
+
```
Please remove the newline.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,154 @@
+	tags   map[string]string
+}
+
+func (e *ElasticsearchQuery) parseSimpleResult(measurement string, searchResult *elastic.SearchResult, acc telegraf.Accumulator) error {
```
As this function cannot error please remove the return parameter to make this obvious to everybody.

srebhan replied:
And can you please make `acc` the first parameter. It's not really a convention but nice to indicate "Hey we pass the acc down here.".

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,154 @@
+
+	// organize the aggregation query data by measurement
+	for _, aggregationQuery := range aggregationQueryList {
+
```
Remove newline. Please only use newlines to indicate new "logical code blocks" as this clutters the code otherwise IMO. :-)

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,154 @@
+			if s, ok := bucket.Key.(string); ok {
+				m.tags[aggName] = s
+			} else {
+				return m, fmt.Errorf("bucket key is not a string")
```
Maybe print the `bucket.Key` to help debugging the problem, but that's only cosmetics.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,154 @@
+
+			// if there are fields present after finishing the bucket, it is a complete metric
+			// store it and clean the fields to start a new metric
+			if len(m.fields) > 0 {
+				acc.AddFields(m.name, m.fields, m.tags)
+				m.fields = make(map[string]interface{})
+			}
```
I think you need a flag to denote that you are at the highest level again before adding the fields. Imagine you go down the tree, say 4 times, then you will add at least 3 times the same metric with a growing number of fields. Remember fields are non-zero in every level except maybe the lowest.
So either add a depth counter and only add the metric if you are at depth zero or split the recursion part from the top-level function...

lpic10 replied:
The metric needs to be added after it reaches the highest level. 

Until it gets there, the tags are being added to the metric, and the fields too, except that the `doc_count` field is overwritten each time it goes up a level to another terms aggregation. Only when it reaches the `metric` aggregation (at the highest level of the recursion) is that it adds other field names like `field_avg` or `field_max`.

So I'm adding the metric when it goes down the first time. By resetting the `fields` and removing that bucket `tag` from the metric struct it means that it will not duplicate that metric when continuing to next bucket (new tag, new field) or going down the recursion. 

As the `metric` struct is returned each time from the higher recursion function down before checking if it needs to be added, normally it will not add again when going down the next time. 

By looking now I need to confirm if it is working correctly when more than one metric field is present.
It seems I will need to add the metric and reset fields at the end of the `AggregationValueMetric` switch case; and the current `addFields` would only be called when no metric field is present.

srebhan replied:
Why don't you collect _all_ fields and _all_ tags during recursion and the call `acc.AddFields(m.name, m.fields, m.tags)` in the measurement loop of `parseAggregationResult()`? I think this will be much easier to understand... I'm still unsure if your description above matches what is done in the code after reading this function multiple times now. :-)

lpic10 replied:
I could do that, so instead of calling `acc.AddFields` directly I could add the resulting metric to a map (or something else) and iterate over that in `parseAggregationResult`, but I think that would be more complex. 

Then I would need to see how to work with the tags slice inside the `metric` struct as I need to add/remove values from it during the recursive calls. As slices work by reference it would change the values in the resulting map as well (I figured that out because I tried that).

I want to add tests to this using JSON responses from these aggregations in Elasticsearch. Maybe by looking at these will help the understanding of this particular part of the code (and maybe to improve with a better solution).

srebhan replied:
To be honest it cannot be as hard to read as this. I _really_ tried to wrap my head around it and I'm still not sure I get all the possible paths, so splitting this into two parts might be a good idea. :-)
I still don't get the idea I think. As far as I get it you get a response that has a tree like structure. Is this correct? That's why you recurse into it and build aggregations on the way. However, you want to collect the whole subtree to form a metric, don't you?
Your code suggest that you call this recursions for _each_ **measurement**, so the first question is, do you expect multiple metrics for each measurement? If yes, do you have a clear signal of when a metric ends and another starts? Currently it looks like you want a metric for each `bucket` and each `metric`, but only at the highest level, all other things further down the tree should be collected/aggregated. Is this correct?

If all of the above is true, moving `acc.AddFields()` to `parseAggregationResult()` is a bad idea. In this case I would suggest to pass a `depth` counter down the recursive function and only call  `acc.AddFields()` if that counter is zero.

lpic10 replied:
> To be honest it cannot be as hard to read as this. I really tried to wrap my head around it and I'm still not sure I get all the possible paths, so splitting this into two parts might be a good idea. :-)

It was painful for me to go back to this code too :)

> I still don't get the idea I think. As far as I get it you get a response that has a tree like structure. Is this correct? 

Yes, exactly, and the objective is somehow to flat that tree.

> That's why you recurse into it and build aggregations on the way. However, you want to collect the whole subtree to form a metric, don't you?

Yes.

> Your code suggest that you call this recursions for _each_ **measurement**, so the first question is, do you expect multiple metrics for each measurement?

Yes, that's the case

> ... do you have a clear signal of when a metric ends and another starts? Currently it looks like you want a metric for each `bucket` and each `metric`, but only at the highest level, all other things further down the tree should be collected/aggregated. Is this correct?

Yes. The metric stops at the highest level as you mention. The highest level is a metric aggregation, when it exists. But a metric aggregation is optional, and the metric may end at the highest level for each bucket.

The metric aggregation (when it exists) is always at the edge of the (sub)tree. 

> If all of the above is true, moving `acc.AddFields()` to `parseAggregationResult()` is a bad idea. In this case I would suggest to pass a `depth` counter down the recursive function and only call `acc.AddFields()` if that counter is zero.

I will investigate this during the weekend if I can, but my feeling is that if the counter is zero it is too late to cal the `acc.AddFields()` as the tags are changing if there are more buckets. The part that I'm really in doubt is about calling it inside the metric aggregation though, I need to test the code using at least 3 terms aggregation and 2 metrics aggregation (the current test data only allowed me 1 metrics aggregation).

srebhan replied:
Thanks for your effort but I think it's worth it... :-)

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_parser.go:
```diff
@@ -0,0 +1,154 @@
+		}
+
+	default:
+		return m, fmt.Errorf("aggregation type returned not supported")
```
Here you might also want to print the unknown aggregation type so we can start debugging from there.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,203 @@
+func (e *ElasticsearchQuery) runAggregationQuery(ctx context.Context, aggregation esAggregation) (*elastic.SearchResult, error) {
+
+	now := time.Now().UTC()
+	from := now.Add(aggregation.QueryPeriod.Duration * -1)
```
Still open...

lpic10 replied:
I noticed I cannot just replace as suggested because `Sub` requires a `Time` as `Add` requires a `Duration`

srebhan replied:
Ok than at least make it
```suggestion
	from := now.Add(-aggregation.QueryPeriod.Duration)
```

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,203 @@
+	now := time.Now().UTC()
+	from := now.Add(aggregation.QueryPeriod.Duration * -1)
+	filterQuery := aggregation.FilterQuery
+
```
Remove newline.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,203 @@
+		}
+	}
+	searchResult, err := search.Do(ctx)
+
```
Remove newline.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,203 @@
+func (e *ElasticsearchQuery) getMetricFields(ctx context.Context, aggregation esAggregation) (map[string]string, error) {
+	mapMetricFields := make(map[string]string)
+
+	if e.esClient == nil {
+		err := e.connectToES()
+		if err != nil {
+			return nil, err
+		}
+	}
```
Do we really need this here? If so you should also care for this in `runAggregationQuery`.

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,203 @@
+		}
+
+		for _, index := range resp {
+			if mappings, ok := index.(map[string]interface{})["mappings"]; ok {
+				if types, ok := mappings.(map[string]interface{}); ok {
+					for _, _type := range types {
+						if field, ok := _type.(map[string]interface{}); ok {
+							for _, _field := range field {
+								if fullname, ok := _field.(map[string]interface{})["full_name"]; ok {
+									if fname, ok := fullname.(string); ok {
+										if mapping, ok := _field.(map[string]interface{})["mapping"]; ok {
+											if fieldType, ok := mapping.(map[string]interface{}); ok {
+												for _, fieldType := range fieldType {
+													if t, ok := fieldType.(map[string]interface{})["type"]; ok {
+														if ftype, ok := t.(string); ok {
+															mapMetricFields[fname] = ftype
+														}
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
```
Please try to invert the logic to save indention levels! Furthermore, please check the type assertions, especially for `index` (first `if`).

lpic10 replied:
Yes, this does not look good... 
I tried reverting at first, but having to initialize each variable before assigning the value after the assertion check wasn't looking good either, unless I'm missing something.

srebhan replied:
Your new solution is really nice!

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query_test.go:
```diff
@@ -0,0 +1,443 @@
+	// populate elasticsearch with nginx_logs test data file
+	file, err := os.Open("testdata/nginx_logs")
+	require.NoError(t, err)
+
```
Remove newline.

srebhan reviewed (commented):
Almost. Please check the recursive function again as I think you add the metrics multiple times...

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+
+[[inputs.elasticsearch_query.aggregation]]
+  ## measurement name for the results of the aggregation query
+  measurement_name = "measurement"
+
+  ## Elasticsearch indexes to query (accept wildcards).
+  index = "index-*"
+
+  ## The date/time field in the Elasticsearch index (mandatory).
+  date_field = "@timestamp"
+
+  ## Time window to query (eg. "1m" to query documents from last minute).
+  ## Normally should be set to same as collection interval
+  query_period = "1m"
+
+  ## Lucene query to filter results
+  filter_query = "*"
+
+  ## Fields to aggregate values (must be numeric fields)
+  metric_fields = ["metric"]
+
+  ## Aggregation function to use on the metric fields
+  ## Valid values are: avg, sum, min, max, sum
+  metric_function = "avg"
+
+  ## Fields to be used as tags
+  ## Must be non-analyzed fields, aggregations are performed per tag
+  tags = ["field.keyword", "field2.keyword"]
+
+  ## Set to true to not ignore documents when the tag(s) above are missing
+  # include_missing_tag = false
+
+  ## String value of the tag when the tag does not exist
+  ## Used when include_missing_tag is true
+  # missing_tag_value = "null"
```
Are all of the non-commented options mandatory and are different to their defaults? If not, please comment as many of them as possible to give the user a sense of what is the bare-minimum to configure.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,277 @@
+		agg.aggregationQueryList = aggregationQueryList
+		e.Aggregations[i] = agg
+	}
+	return err
```
I guess this error can only be `nil`!? If so, please explicitly return `nil`.

srebhan reviewed (commented):
Some minor things remaining besides the big topic of recursion. :-) Can you please remove the commented plugins/lines from `plugins/inputs/elasticsearch_query/dev/telegraf.conf` in order to not add ~5,800 lines of rubbish. :-)

lpic10 commented:
> Some minor things remaining besides the big topic of recursion. :-) Can you please remove the commented plugins/lines from `plugins/inputs/elasticsearch_query/dev/telegraf.conf` in order to not add ~5,800 lines of rubbish. :-)

Well, that dev folder shouldn't be in the commit in the first place (I guess I added all files by mistake at some point). 
I'll remove that on next pull.

Thanks for the review again!

srebhan commented:
@lpic10 thank you for your effort! :+1:

lpic10 commented:
WIP

srebhan commented:
@lpic10 any estimated on when I'll be able to give it another round!?

srebhan commented on plugins/inputs/elasticsearch_query/aggregation_query.go:
```diff
@@ -0,0 +1,215 @@
+		}
+
+		aggregationQueryList = append(aggregationQueryList, aggregationQuery)
+
```
Please change this to make the linter happy.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,304 @@
+import (
+	"context"
+	"fmt"
+	"log"
```
You should _not_ use `log` but rather define 
```golang
Log telegraf.Logger `toml:"-"`
```
in your struct and the use the telegraf logging facility!

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,304 @@
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+
+[[inputs.elasticsearch_query.aggregation]]
+  ## measurement name for the results of the aggregation query
+  measurement_name = "measurement"
+
+  ## Elasticsearch indexes to query (accept wildcards).
+  index = "index-*"
+
+  ## The date/time field in the Elasticsearch index (mandatory).
+  date_field = "@timestamp"
+
+  ## Time window to query (eg. "1m" to query documents from last minute).
+  ## Normally should be set to same as collection interval
+  query_period = "1m"
+
+  ## Lucene query to filter results
+  # filter_query = "*"
+
+  ## Fields to aggregate values (must be numeric fields)
+  # metric_fields = ["metric"]
+
+  ## Aggregation function to use on the metric fields
+  ## Must be set if 'metric_fields' is set
+  ## Valid values are: avg, sum, min, max, sum
+  # metric_function = "avg"
+
+  ## Fields to be used as tags
+  ## Must be text, non-analyzed fields. Metric aggregations are performed per tag
+  # tags = ["field.keyword", "field2.keyword"]
+
+  ## Set to true to not ignore documents when the tag(s) above are missing
+  # include_missing_tag = false
+
+  ## String value of the tag when the tag does not exist
+  ## Used when include_missing_tag is true
+  # missing_tag_value = "null"
```
Please increase the indention of this subtable with two spaces.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,304 @@
+	// retrieve field mapping and build queries only once
+	mapMetricFields, err := e.getMetricFields(ctx, agg)
+	if err != nil {
+		log.Printf("E! %s", err.Error())
+		return
```
Wouldn't it make sense to return this error and handle it in `Init()`? Can you do some sensible processing if this fails?

lpic10 replied:
The idea here is just to wait for the next plugin execution as the data for the query may not be available (yet) on ES, and the plugin init should not fail because of that.

srebhan replied:
Fair enough.

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query_test.go:
```diff
@@ -0,0 +1,682 @@
+	wantQueryResErr           bool
+}
+
+type expectedMetric struct {
+	measurement string
+	fields      map[string]interface{}
+	tags        map[string]string
+}
```
You might want to use `testutil.MustMetric(measurement, fields, tags, time.Unix(0,0)` instead. This way you can later directly test the metrics using `testutil.RequireMetricsEqual(t, expected, acc.GetTelegrafMetrics(), testutil.IgnoreTime())` where expected is `expected := []telegraf.Metric{ testutil.MustMetric(...), testutil.MustMetric(...),...}`

srebhan replied:
Can you please consider this comment.

lpic10 replied:
Done :)

srebhan reviewed (changes requested):
There are still some points you need to address. The most important one is to _not_ use `log` but the telegraf logging factility. See my comments in the code.

srebhan commented:
@lpic10 any chance you work on this?

srebhan commented on plugins/inputs/elasticsearch_query/elasticsearch_query_test.go:
```diff
@@ -0,0 +1,684 @@
+		ResponseTime float64   `json:"response_time"`
+	}
+
+	err := e.connectToES()
+	if err != nil {
+		e.Log.Errorf("error setting up integration test: %s", err.Error())
+		return
+	}
```
Please either simply return the error here and check it in the caller or pass the `testing.T` in here and use `require.NoError()` to fail hard. I prefer the former but I'm also OK with the latter.

srebhan reviewed (commented):
I have some comments for the tests. The other code looks good.

lpic10 commented:
I've changed the tests as requested, let me know if anything else

srebhan commented:
@lpic10 thanks for keep this going! I think the only thing left is the following comment in the tests:

> You might want to use testutil.MustMetric(measurement, fields, tags, time.Unix(0,0) instead. This way you can later directly test the metrics using testutil.RequireMetricsEqual(t, expected, acc.GetTelegrafMetrics(), testutil.IgnoreTime()) where expected is expected := []telegraf.Metric{ testutil.MustMetric(...), testutil.MustMetric(...),...}

 It would be nice if you can care for this one. Everything else looks nice from my point of view.

akrantz01 commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,310 @@
+
+	i, err := strconv.Atoi(esVersionSplit[0])
+	if err != nil || i < 5 {
+		return fmt.Errorf("elasticsearch version %s not supported (currently supported versions are 5.x and 6.x)", esVersion)
```
Should this error on 7.x as well?

lpic10 replied:
Yes, definitely, current code does not work on 7.x versions

akrantz01 reviewed (commented):
Thanks for keeping this PR alive! Overall it looks good; I just left a couple of comments.

lpic10 commented on plugins/inputs/elasticsearch_query/elasticsearch_query.go:
```diff
@@ -0,0 +1,310 @@
+		return fmt.Errorf("elasticsearch urls is not defined")
+	}
+
+	err := e.connectToES()
+	if err != nil {
+		e.Log.Errorf("E! error connecting to elasticsearch: %s", err)
+		return nil
+	}
```
That's the reason I don´t return the error here :)

lpic10 replied:
I don´t know about the Start(), I will see if it makes sense to move some logic there

akrantz01 replied:
Ok, that's what I figured. I think it's fine in the current state since part of the configuration validation depends on a connection to the ES instance.

lpic10 replied:
From what I've checked, Start() applies to service input plugins only. Is that right?

srebhan reviewed (approved):
Thanks @lpic10 for your epic effort in making this fly! Looks good to me.

sspaink reviewed (approved):
Looks great! Awesome work.
